src/libs/ajax/Billing.ts

```diff
@@ -1,15 +1,15 @@
 import _ from 'lodash/fp';
 import * as qs from 'qs';
+import { AzureManagedAppCoordinates } from 'src/billing-logic/AzureManagedAppCoordinates';
+import { BillingProject, BillingProjectMember, BillingRole } from 'src/billing-logic/BillingProject';
+import { GoogleBillingAccount } from 'src/billing-logic/GoogleBillingAccount';
```

Comment:
all of these types appear to be data types returned below from api calls with no or minimal logic.... which makes me think the types should be defined here in Billing.ts (which is what will morph into BillingDataClient).  And I think we might want to re-export them from the logic modules so that layers further up can get the types from there and have no direct dependency to this lower layer.

Thoughts?

Reply:
...here in Billing.ts...  or in a co-located sister file billing-models.ts

Reply:
@msilva-broad But wouldn't they ultimately go into the data provider instead of the client? We aren't going to change the data representation from the Ajax response substantially… there is no compelling reason to, and it would require consumers to change their code. So it would make sense for the return types to exist in the package that contains the data provider definition (which I assumed would be billing-logic) as it defines the contract for the clients.

Reply:
For what it's worth, they can also go in both. The client can define its own interface and the provider can define the type used for Terra UI's data model. They may happen to be the same. Or the provider can re-export from the client.

Reply:
That's true. And at this point, I think it does make more sense for them to live in `Billing.ts`, so I will move them there.

Reply:
Actually, that means`billing-logic` will end up having a dependency on Ajax (or else `billing-ui` will, which is even worse), and I thought is something we want to avoid.

So I'm back to wanting to keep the type definitions in `billing-logic` to reduce dependencies on Ajax/client implementation. I feel like the type definitions are basically the provider API at this point, and I'd rather they stay in `billing-logic` (`billing-core?`).

Reply:
> To make sure I understand the ultimate goal…. would the wrapper class in `src/pages/billing` be the thing that injects the client implementation for the provider? If so, I think that we do want to keep dependencies on the client out of `billing-logic`, and have the client responsible for importing and conforming to the provider models defined in `billing-logic`.

yes, the ultimate goal is to facilitate src/pages/PageX level composition of providers to say "you get your data from this client or clients"... and this page's main component needs the resulting composed provider(s)... as we have been crafting for EnvironementsPage's Environments component.  We compose a hook in some cases, but that hook needs a provider.

It feels like we need a billing-models package to be the facilitator here... src/pages can pull from billing-models to know how to compose things.... data-client and billing-logic providers can do the same to all pull the same contracts/object-shapes.

Reply:
... I am trying to enforce one of the rules of the proposed ui stack:
- lower layers should never reference or be aware of higher layers of the stack

i.e.  a data-client (the lowest layer) should not peek higher up the stack (to provider/logic layer)

Reply:
I think the "client" is intended to be the lowest level of the stack with no imports. The "provider" is like a higher level API client that may shape and/or collect responses into a Terra UI defined data model.

Reply:
I still don't follow though why having a `billing-model` package is any different from putting the models in `billing-logic`.

Reply:
also, provider/logic package can re-export the models to ease package consumption on the consumer side...

Reply:
... but there is a desire to eventually have code-shard data-clients  (at least a core aspect of them)... so that contract tests at the ui level might also be more easily semi-shared accoss ui applications...

...hmmmm...

Reply:
ok... a wrinkle in that path would be that there is overlap in data-client usage...  more then one provider uses certain data-clients... in addition to some providers using more then one data-client... it's many-to-many in the general case 
even if in the majority of cases it's a 1-to-1.

Reply:
I don't see how that is a problem though. The thing that creates the instance of the UI component/page will pass in all the necessary data clients.

What would the alternative be? We don't want providers responsible for picking clients.

Reply:
And even if the clients are in some sort of package to be shared, I don't see why clients need to be at the bottom vs. the logic/provider package.

Reply:
The stack is helpful to visualize things... and it sorta maps to the call-stack of typical key logic flows... I don't think there is a material difference there between thinking of how "central" providers are looking to be, and that the app is the composer of these things... that all aligns.

I think it comes down to a niggling detail at this point... 

If data-client-x is used by two or more providers... where do we define the contract and object-shape types that the data client and the 2+ providers all need?